src/shared/driverUtils.js

```diff
@@ -0,0 +1,258 @@
+            ).toString().trim();
+        }
+        catch (error) {
+            console.debug(`Error starting database container: ${error}.message`);
+            dbStatus.pgReady = false;
+            return dbStatus;
+        }
+    }
+
+    // Loop to check that PostgresDB is ready.
+    console.debug("Checking that PostgresDB is ready...");
+    dbStatus.pgReady = false;
+    for (let i = 0; i < NUM_CHECK_REQUESTS; i++) {
+        console.debug(`... attempt ${i}`);
+        try {
+            execOutput = execSync(
+                `docker exec --user postgres ${container} pg_isready -p ${dbConfig.port}`
+            ).toString().trim();
+        }
+        catch (error) {
+            execOutput = error.message;
+            console.debug(`... error ${execOutput}`);
+        }
+        if (execOutput.indexOf("accepting connections") !== -1) {
+            dbStatus.pgReady = true;
+            break;
+        } else {
+            await fluid.personalData.sleep(DELAY);
+        }
+    }
+    console.debug(`- Attempt to start database, result:  ${dbStatus.pgReady}`);
+    if (dbStatus.pgReady === false) {
+        return dbStatus;    // bail
+    }
+
+    // Container and PostgresDB running, create the actual database to use for
+    // the Postgres tables
+    console.debug(`Creating (or checking existence of) ${dbConfig.database} ...`);
+    try {
+        execOutput = execSync(
+            `docker exec ${container} createdb -p ${dbConfig.port} -U ${dbConfig.user} --echo ${dbConfig.database}`
+        ).toString().trim();
+    }
+    catch (error) {
+        // "Database already exists" not an error in this case.
+        if (error.message.indexOf(`database "${dbConfig.database}" already exists`) !== -1) {
+            dbStatus.pgReady = true;
+        } else {
+            console.debug(`... error ${error.message}`);
+        }
+    }
+    return dbStatus;
+};
+
+/**
+ * Disconnect the database client from the database and stop the database docker
+ * container.  If the container was simply paused at the beginning of testing,
+ * leave it in a paused state, but, if the container was created for testing,
+ * remove it from the container list.
+ *
+ * @param {String} container - Name of the docker container.
+ * @param {Object} wasPaused - Whether the database docker container was paused
+ *                            at the start of testing.
+ * @param {Object} dbRequest - Optional: the PostGresOperations object to
+ *                             disconnect from the database.
+ * @return {Object} result of disconnecting `dbRequest` from the database.
+ */
+fluid.personalData.dockerStopDatabase = async function (container, wasPaused, dbRequest) {
+    console.debug(`- Stopping database container ${container}`);
+    if (dbRequest) {
+        await dbRequest.end();
+    }
+    try {
+        execSync(`docker stop ${container}`);
+        if (!wasPaused) {
+            console.log("WOULD COMPLETELY DESTROY THE DATABASE");
```

Comment:
This is an odd branch - why might we choose to remove the whole container here?

Reply:
> This is an odd branch - why might we choose to remove the whole container here?

@klown, you are probably the better person to answer this question.

Reply:
It appears that this is the only site of the use of the "wasPaused" flag - if we axe this branch, we could axe all the logic associated with the flag

Reply:
@amb26, exactly.  This is left over from when I was thinking of starting a completely new container for creating and initializing a database solely for the tests.  When the tests were done, the whole container could be removed; otherwise you end up with lots of zombie containers.  The motivation was to not pollute the container with the real database with another test database or, fiddle in any way with the real database by isolating the two.  The intended way to distinguish the two cases was via `wasPaused`, but that's iffy.

If it is okay to use one container for both the real database and a test database, then it's not necessary to worry about removing any zombies.